### ISSUE TITLE
Error building SPDK 16.12 NVMeF initiator

### ISSUE THREAD

**Issue by yuliya-navrotskaya** [spdk/spdk#98]
I ran into the following minor error when building SPDK 16.2 NVMeF initiator on the mainline Linux kernel 4.8 with gcc 4.4.7:

nvme_rdma.c:1214: warning: no previous declaration for ‘nvme_rdma_ctrlr_construct’
nvme_rdma.c:1214: error: conflicting types for ‘nvme_rdma_ctrlr_construct’
nvme_rdma.c:1157: note: previous implicit declaration of ‘nvme_rdma_ctrlr_construct’ was here

To resolve this, I've moved nvme_rdma_ctrlr_construct function above nvme_rdma_ctrlr_scan, where it's first called.

**Comment by optimistyzy** [comment]
Hi yuliya-navrotskaya,
Can you list the detailed steps to compile the latest SPDK code?  I do not think that you need to change the source code.

**Comment by yuliya-navrotskaya** [comment]
Hi optimistyzy,

I followed instructions for building https://github.com/spdk/spdk. Made sure I have all prerequisites and downloaded SPDK version 16.12.

cd /path/to/spdk
wget http://fast.dpdk.org/rel/dpdk-16.11.tar.xz
tar xf dpdk-16.11.tar.xz
(cd dpdk-16.11 && make install T=x86_64-native-linuxapp-gcc DESTDIR=.)

Modified CONFIG_RDMA?=y in /path/to/spdk/CONFIG

make DPDK_DIR=./dpdk-16.11/x86_64-native-linuxapp-gcc

Traceback (most recent call last):
  File "scripts/genconfig.py", line 37, in <module>
    print "#undef SPDK_{}".format(var)
ValueError: zero length field name in format
  CC lib/bdev/bdev.o
cc1: error: unrecognized command line option "-Wunused-result"
cc1: error: unrecognized command line option "-Wunused-result"
make[2]: *** [bdev.o] Error 1
make[1]: *** [bdev] Error 2
make: *** [lib] Error 2

Commented out COMMON_CFLAGS += -Wunused-result in /path/to/spdk/mk/spdk.common.mk

make DPDK_DIR=./dpdk-16.11/x86_64-native-linuxapp-gcc
...
  LIB libspdk_util.a
  CC lib/nvme/nvme_ctrlr_cmd.o
  CC lib/nvme/nvme_ctrlr.o
  CC lib/nvme/nvme_ns_cmd.o
  CC lib/nvme/nvme_ns.o
  CC lib/nvme/nvme_pcie.o
  CC lib/nvme/nvme_qpair.o
  CC lib/nvme/nvme.o
  CC lib/nvme/nvme_quirks.o
  CC lib/nvme/nvme_transport.o
  CC lib/nvme/nvme_uevent.o
  CC lib/nvme/nvme_rdma.o
nvme_rdma.c:1047: warning: no previous declaration for ‘nvme_rdma_ctrlr_create_io_qpair’
nvme_rdma.c:1054: warning: no previous declaration for ‘nvme_rdma_ctrlr_enable’
nvme_rdma.c:1139: warning: no previous declaration for ‘nvme_rdma_ctrlr_scan’
nvme_rdma.c: In function ‘nvme_rdma_ctrlr_scan’:
nvme_rdma.c:1157: warning: implicit declaration of function ‘nvme_rdma_ctrlr_construct’
nvme_rdma.c:1157: warning: assignment makes pointer from integer without a cast
nvme_rdma.c: At top level:
nvme_rdma.c:1206: warning: no previous declaration for ‘nvme_rdma_ctrlr_attach’
nvme_rdma.c:1214: warning: no previous declaration for ‘nvme_rdma_ctrlr_construct’
nvme_rdma.c:1214: error: conflicting types for ‘nvme_rdma_ctrlr_construct’
nvme_rdma.c:1157: note: previous implicit declaration of ‘nvme_rdma_ctrlr_construct’ was here
nvme_rdma.c:1257: warning: no previous declaration for ‘nvme_rdma_ctrlr_destruct’
nvme_rdma.c:1271: warning: no previous declaration for ‘nvme_rdma_ctrlr_get_pci_id’
nvme_rdma.c:1277: warning: no previous declaration for ‘nvme_rdma_ctrlr_set_reg_4’
nvme_rdma.c:1283: warning: no previous declaration for ‘nvme_rdma_ctrlr_set_reg_8’
nvme_rdma.c:1289: warning: no previous declaration for ‘nvme_rdma_ctrlr_get_reg_4’
nvme_rdma.c:1302: warning: no previous declaration for ‘nvme_rdma_ctrlr_get_reg_8’
nvme_rdma.c:1308: warning: no previous declaration for ‘nvme_rdma_qpair_submit_request’
nvme_rdma.c:1355: warning: no previous declaration for ‘nvme_rdma_ctrlr_delete_io_qpair’
nvme_rdma.c:1361: warning: no previous declaration for ‘nvme_rdma_ctrlr_reinit_io_qpair’
nvme_rdma.c:1367: warning: no previous declaration for ‘nvme_rdma_qpair_enable’
nvme_rdma.c:1374: warning: no previous declaration for ‘nvme_rdma_qpair_disable’
nvme_rdma.c:1381: warning: no previous declaration for ‘nvme_rdma_qpair_reset’
nvme_rdma.c:1388: warning: no previous declaration for ‘nvme_rdma_qpair_fail’
nvme_rdma.c:1395: warning: no previous declaration for ‘nvme_rdma_qpair_process_completions’
nvme_rdma.c:1478: warning: no previous declaration for ‘nvme_rdma_ctrlr_get_max_xfer_size’
nvme_rdma.c:1485: warning: no previous declaration for ‘nvme_rdma_ctrlr_get_max_io_queue_size’
make[2]: *** [nvme_rdma.o] Error 1
make[1]: *** [nvme] Error 2
make: *** [lib] Error 2

After moving nvme_rdma_ctrlr_construct function above nvme_rdma_ctrlr_scan, where it's first called, in nvme_rdma.c, SPDK built successfully.

**Comment by dverkamp-intel** [comment]
I am unable to reproduce this here:

```
wget https://github.com/spdk/spdk/archive/v16.12.tar.gz
tar xf v16.12.tar.gz
cd spdk-16.12
sed -i s/^CONFIG_RDMA.*/CONFIG_RDMA?=y/ CONFIG
make DPDK_DIR=...
```

The error from genconfig.py indicates that CONFIG has been corrupted somehow; can you try with a clean copy of the source and verify that it works?

**Comment by dverkamp-intel** [comment]
Also, it would be useful to know which compiler version you are using that does not support -Wunused-result - thanks for the report.

**Comment by jimharris** [comment]
It was listed earlier in the thread – gcc 4.4.7.

**Comment by yuliya-navrotskaya** [comment]
Daniel, thanks for trying to reproduce this.

I get the error from genconfig.py with a clean copy of the source. I get this error even if I don't modify CONFIG.  

I used an old version of gcc, 4.4.7, which doesn't support Wunused-result. I switched to gcc version 6.2.1, and that's no longer an issue. However, I still get the same error from nvme_rdma.c.

**Comment by dverkamp-intel** [comment]
I think the actual issue here is the Python version. I didn't realize that positional arguments for string format() required indexes in pre-2.7 versions of Python, as mentioned in https://docs.python.org/2/library/string.html#format-string-syntax

Can you try something like the below diff and verify if it works with your version of python?

```
diff --git a/scripts/genconfig.py b/scripts/genconfig.py
index 4d95691..02f4da9 100755
--- a/scripts/genconfig.py
+++ b/scripts/genconfig.py
@@ -32,9 +32,9 @@ with open('CONFIG') as f:
                     val = args[var]
                 if default.lower() == 'y' or default.lower() == 'n':
                     if val.lower() == 'y':
-                        print "#define SPDK_{} 1".format(var)
+                        print "#define SPDK_{0} 1".format(var)
                     else:
-                        print "#undef SPDK_{}".format(var)
+                        print "#undef SPDK_{0}".format(var)
                 else:
                     strval = val.replace('"', '\"')
-                    print "#define SPDK_{} \"{}\"".format(var, strval)
+                    print "#define SPDK_{0} \"{1}\"".format(var, strval)
```

**Comment by yuliya-navrotskaya** [comment]
Daniel, I verified that it works. I could build SPDK with no issues, and issue #99 went away. Thanks so much for your help!

**Comment by dverkamp-intel** [comment]
I checked in this fix as commit f348a09b507a97af74ddf15bec8e3a66ea908600 - thanks for the report!
